torch/utils/_pytree.py

```diff
@@ -26,194 +28,122 @@
+SUPPORTED_NODES: Dict[Union[Type[Any], str], NodeDef] = {}
+
+# SerializeNodeDef holds two callables:
+# - serialize_fn takes a TreeSpec, and returns a serialized string format of the
+#   context, and the version number
+# - deserialize_fn takes in a string representation of the context, and the
+#   version, and returns the deserialized context
+class SerializeNodeDef(NamedTuple):
+    typ: Type[Any]
+    serialized_type: str
+    to_dumpable_context: Optional[ToDumpableContextFn]
+    from_dumpable_context: Optional[FromDumpableContextFn]
+
+SUPPORTED_SERIALIZED_TYPES: Dict[Union[Type[Any], str], SerializeNodeDef] = {}
+SERIALIZED_TYPE_TO_PYTHON_TYPE: Dict[str, Type[Any]] = {}
 
 def _register_pytree_node(
     typ: Any,
     flatten_fn: FlattenFunc,
     unflatten_fn: UnflattenFunc,
-    to_str_fn: Optional[ToStrFunc] = None,
-    maybe_from_str_fn: Optional[MaybeFromStrFunc] = None,
+    *,
+    to_dumpable_context: Optional[ToDumpableContextFn] = None,
+    from_dumpable_context: Optional[FromDumpableContextFn] = None,
 ) -> None:
-    if to_str_fn is None:
-        def _raise_error(spec: "TreeSpec", child_strings: List[str]) -> str:
-            raise NotImplementedError(f"Serializing {typ} not implemented")
-        to_str_fn = _raise_error
-
-    if maybe_from_str_fn is None:
-        def dummy_to_str(str_spec: str) -> Optional[Tuple[Any, Context, str]]:
-            return None
-        maybe_from_str_fn = dummy_to_str
-
-    assert to_str_fn is not None
-    assert maybe_from_str_fn is not None
-    node_def = NodeDef(typ, flatten_fn, unflatten_fn, to_str_fn, maybe_from_str_fn)
+    node_def = NodeDef(
+        typ,
+        flatten_fn,
+        unflatten_fn,
+    )
     SUPPORTED_NODES[typ] = node_def
 
-def _str_to_dict(str_spec: str) -> Tuple[List[str], str]:
-    assert str_spec[1] == "("
-    assert str_spec[-1] == ")"
-    context_and_child_strings = str_spec[2:-1]
-
-    child_strings = []
-    context_strings = []
-    nested_parentheses = 0
-    start_index = 0
-    for i, char in enumerate(context_and_child_strings):
-        if char == ":":
-            if nested_parentheses == 0:
-                context_strings.append(context_and_child_strings[start_index:i])
-                start_index = i + 1
-        elif char == "(":
-            nested_parentheses += 1
-        elif char == ")":
-            nested_parentheses -= 1
-
-        if nested_parentheses == 0 and char == ",":
-            child_strings.append(context_and_child_strings[start_index:i])
-            start_index = i + 1
-
-    child_strings.append(context_and_child_strings[start_index:])
-    return context_strings, ','.join(child_strings)
+    serialized_type = f"{typ.__module__}.{typ.__name__}"
+    serialize_node_def = SerializeNodeDef(
+        typ, serialized_type, to_dumpable_context, from_dumpable_context
+    )
+    SUPPORTED_SERIALIZED_TYPES[typ] = serialize_node_def
+    SERIALIZED_TYPE_TO_PYTHON_TYPE[serialized_type] = typ
+
```

Comment:
nit: should validate that either both or none of {to_dumpable_context, from_dumpable_context} are passed